climada_petals/engine/supplychain.py

```diff
@@ -99,28 +100,49 @@ def calc_B(Z, x):
     This function adapts pymrio.tools.iomath.calc_A to compute
     the allocation coefficients matrix B.
     """
+    # Convert x to a NumPy array
+    x = np.asarray(x)
 
-    if isinstance(x, (pd.DataFrame, pd.Series)):
-        x = x.to_numpy()
-    if not isinstance(x, np.ndarray) and (x == 0):
-        recix = 0
-    else:
-        with warnings.catch_warnings():
-            # Ignore devide by zero warning, we set to 0 afterwards
-            warnings.filterwarnings("ignore", message="divide by zero")
-            recix = 1 / x
-        recix[np.isinf(recix)] = 0
+    # Handle zero values in x
+    with np.errstate(divide='ignore'):
+        recix = np.where(x == 0, 0, 1 / x).reshape((1, -1))
 
+    # Calculate B matrix
     if isinstance(Z, pd.DataFrame):
-        return pd.DataFrame(Z.to_numpy() * recix, index=Z.index, columns=Z.columns)
+        return pd.DataFrame(np.transpose(Z.values) * recix, index=Z.index, columns=Z.columns)
+    else:
+        return np.transpose(Z) * recix
 
-    return Z * recix
 
+def calc_G(B):
+    """Calculate the Ghosh inverse matrix G either from B
+    G = inverse matrix of (I - B) where I is an identity matrix of same shape as B.
+    Note that we define G as the transpose of the Ghosh inverse matrix, so that we can apply the factors of
+    production intensities from the left-hand-side for both Leontief and Ghosh attribution. In this way the
+    multipliers have the same (vector) dimensions and can be added.
+    Parameters
+    ----------
+    B : pandas.DataFrame or numpy.array
+        Symmetric input output table (coefficients)
+    Returns
+    -------
+    pandas.DataFrame or numpy.array
+        Ghosh input output table G
+        The type is determined by the type of B.
+        If DataFrame index/columns as B
+    """
+    I = np.eye(B.shape[0])
+    if isinstance(B, pd.DataFrame):
+        return pd.DataFrame(
+            np.linalg.inv(I - B), index=B.index, columns=B.columns
```

Comment:
Shouldn't inverting switch the index labels with column labels? Maybe they are identical here?

Reply:
Yes they are identical